File: app/lib/agents/deep-research/runner.ts
/**
 * Programmatic runner for the deep research agent.
 *
 * Uses ADK's InMemoryRunner to execute the multi-squad SequentialAgent pipeline.
 * Supports both batch and streaming modes.
 */

import { InMemoryRunner } from "@google/adk";
import { createDeepResearchAgent, type AgentConfig } from "./agent";
import { formatConversationHistory } from "@/app/lib/types";

export interface DeepResearchResult {
  answer: string;
  enrichedQuery: string;
  researchQuestions: string;
  sectionFindings: string;
}

/** Maps session state keys to their corresponding result field names. */
const STATE_KEY_MAP: Record<string, keyof DeepResearchResult> = {
  enriched_query: "enrichedQuery",
  research_questions: "researchQuestions",
  final_report: "answer",
};

/** A pipeline step event (query analyzer, question expander, report). */
export interface StepEvent {
  type: "step";
  field: keyof DeepResearchResult;
  value: string;
}

/** Emitted when we know how many researchers will run. */
export interface ResearchersInitEvent {
  type: "researchers_init";
  count: number;
  labels: string[];
}

/** Emitted when an individual researcher completes. */
export interface ResearcherDoneEvent {
  type: "researcher_done";
  index: number;
  label: string;
  value: string;
}

/** Log entry emitted for every ADK event with agent name and token usage. */
export interface LogEvent {
  type: "log";
  agent: string;
  message: string;
  promptTokens: number;
  responseTokens: number;
  totalTokens: number;
  timestamp: number;
  researcherIndex?: number;
}

export type PipelineEvent = StepEvent | ResearchersInitEvent | ResearcherDoneEvent | LogEvent;

/**
 * Streaming runner — yields PipelineEvent objects as each pipeline step writes
 * its output to session state.
 */
export async function* streamDeepResearch(
  query: string,
  config: AgentConfig
): AsyncGenerator<PipelineEvent, void, void> {
  const agent = createDeepResearchAgent(config);
  const runner = new InMemoryRunner({ agent });

  const historyPrefix = formatConversationHistory(config.conversationHistory ?? []);
  const message = historyPrefix + query;

  let researcherLabels: string[] = [];

  for await (const event of runner.runEphemeral({
    userId: "research-user",
    newMessage: {
      role: "user",
      parts: [{ text: message }],
    },
  })) {
    // Emit a log event for every ADK event that has an author
    const author = event.author ?? "unknown";
    const usage = event.usageMetadata;
    const promptTokens = usage?.promptTokenCount ?? 0;
    const responseTokens = usage?.candidatesTokenCount ?? 0;
    const totalTokens = usage?.totalTokenCount ?? (promptTokens + responseTokens);

    if (author !== "user") {
      const stateKeys = event.actions?.stateDelta ? Object.keys(event.actions.stateDelta) : [];
      // Skip events whose state deltas are handled specially below (researcher_log, researcher_N, etc.)
      const HANDLED_KEY_RE = /^(researcher_log|researcher_count|researcher_labels|researcher_\d+)$/;
      const unhandledKeys = stateKeys.filter((k) => !HANDLED_KEY_RE.test(k));
      const detail = unhandledKeys.length > 0 ? ` → wrote [${unhandledKeys.join(", ")}]` : "";
      // Skip noise events: no tokens, no meaningful state writes
      if (totalTokens > 0 || unhandledKeys.length > 0) {
        yield {
          type: "log" as const,
          agent: author,
          message: `${author}${detail}`,
          promptTokens,
          responseTokens,
          totalTokens,
          timestamp: Date.now(),
        };
      }
    }

    const delta = event.actions?.stateDelta;
    if (!delta) continue;

    for (const [key, value] of Object.entries(delta)) {
      if (typeof value !== "string") continue;

      // Researcher metadata: labels and count
      if (key === "researcher_labels") {
        try {
          researcherLabels = JSON.parse(value);
        } catch {
          researcherLabels = [];
        }
        const count =
          delta["researcher_count"] != null
            ? Number(delta["researcher_count"])
            : researcherLabels.length;
        yield {
          type: "researchers_init",
          count,
          labels: researcherLabels,
        };
        continue;
      }

      // Skip the count key (already handled above)
      if (key === "researcher_count") continue;

      // Researcher log events from DynamicResearchSquad
      if (key === "researcher_log") {
        try {
          const log = JSON.parse(value);
          yield {
            type: "log" as const,
            agent: log.agent ?? "researcher",
            message: log.message ?? "",
            promptTokens: log.promptTokens ?? 0,
            responseTokens: log.responseTokens ?? 0,
            totalTokens: log.totalTokens ?? 0,
            timestamp: log.timestamp ?? Date.now(),
            researcherIndex: typeof log.researcherIndex === "number" ? log.researcherIndex : undefined,
          };
        } catch {
          // ignore malformed log
        }
        continue;
      }

      // Individual researcher completion: researcher_0, researcher_1, etc.
      const researcherMatch = key.match(/^researcher_(\d+)$/);
      if (researcherMatch) {
        const index = Number(researcherMatch[1]);
        yield {
          type: "researcher_done",
          index,
          label: researcherLabels[index] ?? `Researcher ${index + 1}`,
          value,
        };
        continue;
      }

      // Standard pipeline step
      if (key in STATE_KEY_MAP) {
        yield {
          type: "step",
          field: STATE_KEY_MAP[key],
          value,
        };
      }
    }
  }
}

/**
 * Batch runner — runs the full pipeline and returns collected results.
 */
export async function runDeepResearch(
  query: string,
  config: AgentConfig
): Promise<DeepResearchResult> {
  const result: DeepResearchResult = {
    answer: "(no report produced)",
    enrichedQuery: "",
    researchQuestions: "",
    sectionFindings: "",
  };

  for await (const event of streamDeepResearch(query, config)) {
    if (event.type === "step") {
      result[event.field] = event.value;
    }
  }

  return result;
}

File: fastapi/agents/deep_research/runner.py
"""
Programmatic runner for the deep research agent.

Uses ADK's InMemoryRunner to execute the multi-squad SequentialAgent pipeline.
Yields PipelineEvent dicts for SSE streaming.
"""

from __future__ import annotations

import json
import re
import time
from typing import AsyncGenerator, Literal, TypedDict

from models import ConversationMessage, format_conversation_history
from agents.runner_helper import run_agent_ephemeral, extract_usage_tokens
from agents.deep_research.pipeline import create_deep_research_agent


# ---------------------------------------------------------------------------
# Event types (match the TypeScript originals exactly)
# ---------------------------------------------------------------------------

class StepEvent(TypedDict):
    type: Literal["step"]
    field: str  # "enrichedQuery" | "researchQuestions" | "answer"
    value: str


class ResearchersInitEvent(TypedDict):
    type: Literal["researchers_init"]
    count: int
    labels: list[str]


class ResearcherDoneEvent(TypedDict):
    type: Literal["researcher_done"]
    index: int
    label: str
    value: str


class LogEvent(TypedDict):
    type: Literal["log"]
    agent: str
    message: str
    promptTokens: int
    responseTokens: int
    totalTokens: int
    timestamp: int
    researcherIndex: int | None


from typing import Union
PipelineEvent = Union[StepEvent, ResearchersInitEvent, ResearcherDoneEvent, LogEvent]


# Maps session state keys to SSE field names
STATE_KEY_MAP: dict[str, str] = {
    "enriched_query": "enrichedQuery",
    "research_questions": "researchQuestions",
    "final_report": "answer",
}

# Keys handled specially (not included in generic log detail)
HANDLED_KEY_RE = re.compile(r"^(researcher_log|researcher_count|researcher_labels|researcher_\d+)$")


async def stream_deep_research(
    query: str,
    *,
    project: str,
    location: str,
    fast_model: str,
    advanced_model: str,
    report_model: str,
    rag_corpus: str,
    conversation_history: list[ConversationMessage] | None = None,
) -> AsyncGenerator[PipelineEvent, None]:
    """
    Streaming runner — yields PipelineEvent dicts as each pipeline step
    writes its output to session state.
    """
    agent = create_deep_research_agent(
        fast_model=fast_model,
        advanced_model=advanced_model,
        report_model=report_model,
        project=project,
        location=location,
        rag_corpus=rag_corpus,
    )

    history_prefix = format_conversation_history(conversation_history or [])
    message = history_prefix + query

    researcher_labels: list[str] = []

    async for event in run_agent_ephemeral(agent, message, user_id="research-user", app_name="deep_research_app"):
        author = event.author or "unknown"
        prompt_tokens, response_tokens, total_tokens = extract_usage_tokens(event)

        if author != "user":
            state_delta = {}
            if hasattr(event, "actions") and event.actions:
                state_delta = getattr(event.actions, "state_delta", {}) or {}

            state_keys = list(state_delta.keys()) if state_delta else []
            unhandled_keys = [k for k in state_keys if not HANDLED_KEY_RE.match(k)]
            detail = f" → wrote [{', '.join(unhandled_keys)}]" if unhandled_keys else ""

            # Emit log event (skip noise: no tokens and no meaningful state writes)
            if total_tokens > 0 or unhandled_keys:
                yield LogEvent(
                    type="log",
                    agent=author,
                    message=f"{author}{detail}",
                    promptTokens=prompt_tokens,
                    responseTokens=response_tokens,
                    totalTokens=total_tokens,
                    timestamp=int(time.time() * 1000),
                    researcherIndex=None,
                )

        # Process state delta
        state_delta = {}
        if hasattr(event, "actions") and event.actions:
            state_delta = getattr(event.actions, "state_delta", {}) or {}

        if not state_delta:
            continue

        for key, value in state_delta.items():
            if not isinstance(value, str):
                continue

            # Researcher metadata: labels and count
            if key == "researcher_labels":
                try:
                    researcher_labels = json.loads(value)
                except (json.JSONDecodeError, TypeError):
                    researcher_labels = []
                count_val = state_delta.get("researcher_count")
                count = int(count_val) if count_val is not None else len(researcher_labels)
                yield ResearchersInitEvent(
                    type="researchers_init",
                    count=count,
                    labels=researcher_labels,
                )
                continue

            if key == "researcher_count":
                continue

            # Researcher log events
            if key == "researcher_log":
                try:
                    log = json.loads(value)
                    yield LogEvent(
                        type="log",
                        agent=log.get("agent", "researcher"),
                        message=log.get("message", ""),
                        promptTokens=log.get("promptTokens", 0),
                        responseTokens=log.get("responseTokens", 0),
                        totalTokens=log.get("totalTokens", 0),
                        timestamp=log.get("timestamp", int(time.time() * 1000)),
                        researcherIndex=log.get("researcherIndex"),
                    )
                except (json.JSONDecodeError, TypeError):
                    pass
                continue

            # Individual researcher completion
            researcher_match = re.match(r"^researcher_(\d+)$", key)
            if researcher_match:
                index = int(researcher_match.group(1))
                yield ResearcherDoneEvent(
                    type="researcher_done",
                    index=index,
                    label=researcher_labels[index] if index < len(researcher_labels) else f"Researcher {index + 1}",
                    value=value,
                )
                continue

            # Standard pipeline step
            if key in STATE_KEY_MAP:
                yield StepEvent(
                    type="step",
                    field=STATE_KEY_MAP[key],
                    value=value,
                )
